File: Python3/476. Number Complement.py
class Solution:
    def findComplement(self, num):
        """
        :type num: int
        :rtype: int
        """
        # count = 0 # 本人拙劣的解法
        # res = num
        # while res:
        #     res //= 2
        #     count += 1
        # return 2**count-1-num
        
        mask = ~0 # 位运算，速度比上面这个快4ms (36ms vs 40ms)
        while mask & num:
            mask <<= 1
        return ~mask & ~num

File: Golang/476. Number Complement.go
// func findComplement(num int) int {
//     var count, total, op int
//     for num > 0 {
//         if num%2 > 0 {
//             op = 0
//         } else {
//             op = 1
//         }
//         total += pow2(count) * op
//         count++
//         num /= 2
//     }
//     return total
// }

// func pow2(count int) int {
//     if count == 0 {
//         return 1
//     } else {
//         return 2*pow2(count-1)
//     }
// }

func findComplement(num int) int {
    var mask int = ^0
    for mask&num > 0 {
        mask <<= 1
    }
    return ^mask & ^num
}
